src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts

```diff
@@ -0,0 +1,504 @@
+		const decorations: IModelDeltaDecoration[] = [];
+		for (const range of ranges) {
+			const ranges = UnicodeTextModelHighlighter.computeUnicodeHighlights(this._model, this._options, range);
+			for (const range of ranges) {
+				decorations.push({ range, options: this._options.includeComments ? DECORATION : DECORATION_HIDE_IN_COMMENTS });
+			}
+		}
+		this._decorationIds = new Set(this._editor.deltaDecorations([...this._decorationIds], decorations));
+	}
+
+	public getDecorationInfo(decorationId: string): UnicodeHighlighterDecorationInfo | null {
+		if (!this._decorationIds.has(decorationId)) {
+			return null;
+		}
+		const range = this._editor.getModel().getDecorationRange(decorationId)!;
+		const text = this._editor.getModel().getValueInRange(range);
+		return {
+			reason: computeReason(text, this._options)!,
+		};
+	}
+}
+
+export class UnicodeHighlighterHover implements IHoverPart {
+	constructor(
+		public readonly owner: IEditorHoverParticipant<UnicodeHighlighterHover>,
+		public readonly range: Range,
+		public readonly decoration: IModelDecoration
+	) { }
+
+	public isValidForHoverAnchor(anchor: HoverAnchor): boolean {
+		return (
+			anchor.type === HoverAnchorType.Range
+			&& this.range.startColumn <= anchor.range.startColumn
+			&& this.range.endColumn >= anchor.range.endColumn
+		);
+	}
+}
+
+export class UnicodeHighlighterHoverParticipant implements IEditorHoverParticipant<MarkdownHover> {
+
+	constructor(
+		private readonly _editor: ICodeEditor,
+		private readonly _hover: IEditorHover,
+		@IModeService private readonly _modeService: IModeService,
+		@IOpenerService private readonly _openerService: IOpenerService,
+	) {
+	}
+
+	computeSync(anchor: HoverAnchor, lineDecorations: IModelDecoration[]): MarkdownHover[] {
+		if (!this._editor.hasModel() || anchor.type !== HoverAnchorType.Range) {
+			return [];
+		}
+
+		const model = this._editor.getModel();
+
+		const unicodeHighlighter = this._editor.getContribution<UnicodeHighlighter>(UnicodeHighlighter.ID);
+
+
+		const result: MarkdownHover[] = [];
+		for (const d of lineDecorations) {
+
+			const highlightInfo = unicodeHighlighter.getDecorationInfo(d.id);
+			if (!highlightInfo) {
+				continue;
+			}
+			const char = model.getValueInRange(d.range);
+			// text refers to a single character.
+			const codePoint = char.codePointAt(0)!;
+
+			function formatCodePoint(codePoint: number) {
+				return `\`U+${codePoint.toString(16)}\` "${`${escapedMarkdownInlineCode(String.fromCodePoint(codePoint))}`}"`;
+			}
+
+			const codePointStr = formatCodePoint(codePoint);
+
+			let reason: string;
+			switch (highlightInfo.reason.kind) {
+				case 'ambiguous':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsAmbiguous',
+						'The character {0} could be confused with the more common character {1}.',
+						codePointStr,
+						formatCodePoint(highlightInfo.reason.confusableWith.codePointAt(0)!)
+					);
+					break;
+
+				case 'invisible':
+					reason = nls.localize(
```

Comment:
Bi-Di Control characters should never be really rendered in the hover:

![image](https://user-images.githubusercontent.com/5047891/142647142-606ef0f8-2f63-493b-a02c-41d2a491ada6.png)

Reply:
I think it does not make sense to render invisible characters anyway. I will just remove the `"..."`.